### PR TITLE
Stand alone script to run a single atm 3DVar analysis

### PR DESCRIPTION
**Pull request by CoryMartin-NOAA**
This PR adds a new ush script, `run_single_atm_var_analysis.py`.

This script does the following:

- Reads an input user configurable YAML (see below)
- Creates a config dict matching what `ush/genYAML` expects
- Creates the 3DVar YAML
- Uses R2D2 to stage backgrounds, obs, bias coeff files
- Links fix files from a fix/ directory I compiled
- Links the executable to the working directory
- Sets up all additionally needed directories
- Generates a SLURM batch script and submits the job

Along with this script, the associated functions have been added/modified to support each of the above steps.

A sample input YAML:
```
working directory: /work2/noaa/stmp/cmartin/gdas_single_test
GDASApp home: /work2/noaa/da/cmartin/GDASApp/work/GDASApp
analysis options:
  berror_yaml: /work2/noaa/da/cmartin/GDASApp/work/GDASApp/parm/atm/berror/staticb_bump.yaml
  obs_yaml_dir: /work2/noaa/da/cmartin/GDASApp/work/GDASApp/parm/atm/obs/config
  var_yaml: /work2/noaa/da/cmartin/GDASApp/work/GDASApp/parm/atm/variational/3dvar_dripcg.yaml
  obs_list: /work2/noaa/da/cmartin/GDASApp/work/GDASApp/parm/atm/obs/lists/gdas_prototype.yaml
  gdas_fix_root: /work2/noaa/da/cmartin/GDASApp/fix
  atm: true
  layout_x: 1
  layout_y: 1
  atm_window_length: PT6H
  valid_time: 2021-12-21T06:00:00Z
  dump: gdas
  case: C96
  case_enkf: C48
  dohybvar: true
  levs: 128
job options:
  machine: orion
  account: da-cpu
  queue: debug
  partition: debug
  walltime: '10:00'
  ntasks: 6
  cpus-per-task: 1
  modulepath: /work2/noaa/da/cmartin/GDASApp/work/GDASApp/modulefiles
```

To run:
```
module use /path/to/clone/modulefiles
module load GDAS/orion
export PYTHONPATH:/path/to/clone/ush:$PYTHONPATH
python run_single_atm_var_analysis.py --config sample.yaml
```

Sample working directory can be found at: `/work2/noaa/stmp/cmartin/gdas_single_test`

### CODE REVIEW

**Comment by CoryMartin-NOAA** [comment]
Note `dohybvar=true` isn't quite correct, as it is not hybrid-3DVar, but this controls making the analysis resolution 1/2 that of the background resolution.

@FanHan-NOAA you might want to try being a beta tester here to clone, build, and run this single script, as it will help with your interpolation work for the ISDA conference.

File: ush/ufsda/misc_utils.py
Comment (aerorahul) [inline ush/ufsda/misc_utils.py]: I would suggest using `scheduler` instead of `machine` where these kind of decisions need to be made.

Comment (CoryMartin-NOAA) [reply ush/ufsda/misc_utils.py]: Yeah I thought about that, but wanted to make it simpler for the user. Should I do the workflow approach where scheduler is the if statement but it is determined based on a dict (`if scheduler[job_config['machine']] == 'slurm'`)?

**Comment by FanHan-NOAA** [comment]
> Note `dohybvar=true` isn't quite correct, as it is not hybrid-3DVar, but this controls making the analysis resolution 1/2 that of the background resolution.
> 
> @FanHan-NOAA you might want to try being a beta tester here to clone, build, and run this single script, as it will help with your interpolation work for the ISDA conference.

Great! Thanks for this addition. I'll start testing it soon.

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA :  As we review this PR shall we ignore, for the the time being, pieces specific to Hera and focus solely on Orion?

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA I'm open to see what you and others think. Currently this will *only* work on Orion, as I need to stage fix files and a R2D2 database on Hera. Does it make sense to support both systems now, or focus on Orion and port when things have stabilized slightly?

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA, I agree with you.  To keep things moving focus on Orion and port when GDASApp has stabilized.

File: modulefiles/GDAS/orion.lua
Comment (RussTreadon-NOAA) [inline modulefiles/GDAS/orion.lua:56]: Not a request for a change, but echoing previous comments expressing need for a central location (e.g., role account) in which to store shared files.

File: parm/atm/berror/staticb_identity.yaml
Comment (RussTreadon-NOAA) [inline parm/atm/berror/staticb_identity.yaml:1]: Just curious.   Is `FV3JEDI-ID` a standard JEDI variable or name for the identity model?   I can't find `FV3JEDI-ID` defined in `feature/single_anl`

Comment (CoryMartin-NOAA) [reply parm/atm/berror/staticb_identity.yaml:1]: This is the FV3-JEDI YAML entry for the identity **B**, see for example: https://github.com/JCSDA-internal/fv3-jedi/blob/7e9e61222e32e494482dd442d6e86d9d31098c81/test/testinput/hyb-3dvar_geos.yaml#L29

File: parm/atm/obs/config/amsua_n19.yaml
Comment (RussTreadon-NOAA) [inline parm/atm/obs/config/amsua_n19.yaml:47]: This comment is not for the output satbias_cov file but rather for the input file.   I think the input covariance file should be `$(BIAS_DIR)/$(BIAS_PREFIX).amsua_n19.satbias_cov.$(BIAS_DATE).nc4`

Comment (CoryMartin-NOAA) [reply parm/atm/obs/config/amsua_n19.yaml:47]: Yes, I opened an issue here https://github.com/JCSDA-internal/ufo/issues/1964 I am not sure about if they are supposed to be separate or not, as the converter suggests they are all in one file...

File: parm/atm/obs/config/sondes.yaml
Comment (RussTreadon-NOAA) [inline parm/atm/obs/config/sondes.yaml:6]: Removing `LaunchTime` is only necessary because this variable is not in the R2D2 obs.  If we regenerate the R2D2 obs with this variable, we should add `LaunchTime` back.  Is this correct?   If "yes", what's the long term plan for sonde data?  Will it include `LanuchTime`?   If this is where we are heading, should we redo the R2D2 database to include this varaible now?

File: test/test_generate_yaml.py
Comment (RussTreadon-NOAA) [inline test/test_generate_yaml.py:13]: If users want to use `staticb_identity.yaml`, is this the line they need to change?

File: parm/atm/obs/config/sondes.yaml
Comment (CoryMartin-NOAA) [reply parm/atm/obs/config/sondes.yaml:6]: Yes, this is because currently obs come from GSI ncdiags. The operational ncdiags do not contain `LaunchTime`. To get it from ncdiags, we would have to rerun the GSI observer with a modified GSI executable. Long term, this should be coming directly from BUFR files, which should have (hopefully) `LaunchTime`.

File: test/genYAML_prep.sh
Comment (RussTreadon-NOAA) [inline test/genYAML_prep.sh:9]: If users want to run with `staticb_identity.yaml`, is this the line they need to change?

File: modulefiles/GDAS/orion.lua
Comment (CoryMartin-NOAA) [reply modulefiles/GDAS/orion.lua:56]: Yes, tagging @dtkleist for awareness of a use case

**Review by RussTreadon-NOAA — COMMENTED** [review]
Successfully installed and ran `run_single_atm_var_analysis.py` on Orion.   The description for PR #42 provides a `sample.yaml`.   Should we commit a templated `sample.yaml` to  `feature/single_anl`?

File: test/test_generate_yaml.py
Comment (CoryMartin-NOAA) [reply test/test_generate_yaml.py:13]: Correct.

**Comment by CoryMartin-NOAA** [comment]
> Should we commit a templated `sample.yaml` to `feature/single_anl`?

I was actually thinking the next step is to add a "regression test" that would do a nccmp on the increment file and compare it to the existing increment file and determine if the analysis still is produced. That way the regression test would hold the sample YAML file, but I can alternatively create a `parm/examples` directory or something like that?

File: test/test_generate_yaml.py
Comment (CoryMartin-NOAA) [reply test/test_generate_yaml.py:13]: This is for the unit test. Most users would change it in `sample.yaml`.

File: parm/atm/berror/staticb_identity.yaml
Comment (RussTreadon-NOAA) [reply parm/atm/berror/staticb_identity.yaml:1]: Thanks, @CoryMartin-NOAA .   There's still much for me to learn.

File: parm/atm/obs/config/amsua_n19.yaml
Comment (RussTreadon-NOAA) [reply parm/atm/obs/config/amsua_n19.yaml:47]: OK.   Good to open [ufo #1964](https://github.com/JCSDA-internal/ufo/issues/1964) to get clarification.

File: parm/atm/obs/config/sondes.yaml
Comment (RussTreadon-NOAA) [reply parm/atm/obs/config/sondes.yaml:6]: Got it.  It would be nice to directly read from BUFR files sooner rather than later.   Populating R2D2 with GSI ncdiags is extra work and consumes additional disk space.

**Comment by RussTreadon-NOAA** [comment]
> > Should we commit a templated `sample.yaml` to `feature/single_anl`?
> 
> I was actually thinking the next step is to add a "regression test" that would do a nccmp on the increment file and compare it to the existing increment file and determine if the analysis still is produced. That way the regression test would hold the sample YAML file, but I can alternatively create a `parm/examples` directory or something like that?

What you outline is feasible.   This will provide users with an example of what `sample.yaml` looks like.

**Review by RussTreadon-NOAA — DISMISSED** [review]
Thank you, @CoryMartin-NOAA , for answering my questions.   The changes look good to me. 
I see that you are working with @aerorahul regarding `scheduler` in `ush/ufsda/misc_utils.py`.   I'm more an observer than a reviewer when it comes to these details.  It's a good learning experience for me.  Thanks!

File: ush/ufsda/misc_utils.py
Comment (CoryMartin-NOAA) [reply ush/ufsda/misc_utils.py]: @aerorahul how is this?

**Comment by CoryMartin-NOAA** [comment]
Thanks @RussTreadon-NOAA , I have pushed a new commit that dismissed your approval, if @aerorahul is happy with my new changes, I will merge this in once I have one 'approval'

File: ush/run_single_atm_var_analysis.py
Comment (aerorahul) [inline ush/run_single_atm_var_analysis.py:31]: What's going on here? And why is this import in a function?

**Review by aerorahul — DISMISSED** [review]
Changes look good.

**Review by RussTreadon-NOAA — APPROVED** [review]
Looks good to me.

**Pull request merged**
